### ISSUE TITLE
Feature Request: CLI build static site

### ISSUE THREAD

**Issue by stursby** [vapid/vapid#9]
Would it be possible to add something like `vapid build .` to generate the static site? Currently, it looks like you can only deploy via the Vapid hosting.

I see there's a [Builder Class](https://github.com/vapid/vapid/blob/9937092ed98609d7e4532c62c774ea7ef2231646/lib/builder.js) but not sure how to implement it. Any ideas/thoughts would be much appreciated.

Thanks for the cool project!

**Comment by stursby** [comment]
Is something like this going in the right direction? Or should I be looking at `Generate` instead of `Build`?? 🤔

```javascript
const Builder = require('./node_modules/vapid-cli/lib/builder.js')
const { resolve } = require('path')

const templatesDir = resolve(__dirname, 'www')

async function build() {
  const builder = new Builder({ templatesDir })
  try {
    await builder.init()
    const files = await builder.build()
  } catch (err) {
    console.log(err)
  } 
}

build()
```

**Comment by trst** [comment]
It would be doubly awesome to have a way to generate a static site from the dashboard interface as well.

**Comment by KeizerDev** [comment]
This is for me a MUST to use it in production. I want to deploy it to my vps.

**Comment by srobbin** [comment]
@stursby:  Thanks for bringing this up. Answers inline (slightly reordered):

> Currently, it looks like you can only deploy via the Vapid hosting.

True, `vapid deploy` posts to the Vapid hosting service, but you can definitely host Vapid elsewhere, like Heroku or any service that supports Node.js. So in that sense, you can deploy anywhere.

Would it be helpful if Vapid let you override `vapid deploy` so that it acted as some sort of alias to an NPM script? e.g. If `vapid deploy` kicked off `git push heroku master`?

> Would it be possible to add something like vapid build . to generate the static site?

I'm curious to hear more about your use case (and from @trst and @KeizerDev too). Are you planning to use Vapid's dashboard to edit content on your production server? What is appealing about using Vapid over one of the existing static site generators, like Jekyll or Middleman?

> I see there's a Builder Class but not sure how to implement it. Any ideas/thoughts would be much appreciated. Getting the following Error: TypeError: Cannot read property 'rebuild' of undefined

The error is happening because [Builder is looking for the Section model](https://github.com/vapid/vapid/blob/master/lib/builder.js#L43), which gets [passed into the constructor](https://github.com/vapid/vapid/blob/master/lib/builder.js#L19). In any case, I'm not sure that Builder is isolated enough to be used as a standalone class, like your example.

Let's chat about the first two points, and then hopefully I can shed some more light on the third.

**Comment by KeizerDev** [comment]
>I'm curious to hear more about your use case (and from @trst and @KeizerDev too). Are you planning to use Vapid's dashboard to edit content on your production server? What is appealing about using Vapid over one of the existing static site generators, like Jekyll or Middleman?

I would like to use the Vapid dashboard locally to maintain multiple sites and when I'm done I would deploy them on my vps like you would do with Jekyll. What I like about Vapid instead of Jekyll (I have no experience with middleman so can't say anything about that) is the way of just declaring a variable and then fill them with your data instead of the otherway around like you do in Jekyll. Same for articles (repeating content), it's a cleaner and more flexible way of working imho.

I would like to have some command like `vapid build` to just generate static html/css so I can serve it with apache/nginx and not have to worry about anything else like hosting an admin panel or again another service to hosts my sites on (i'd like to keep my own control over what I host). 

Maybe this is not the way you designed it or not the way you liked it to be, this is just how it would suit for me and how I would liked it to be. 

Let me know what you think @srobbin

**Comment by trst** [comment]
If it were just for me, I might stick to a static generator, like Jekyll, however, I think Vapid really does benefit a client with less technical savvy than a developer. I would like to give clients an easy to update solution with the _option_ to build statically for security purposes. These static files could be stored in a git repository through a cron job (or whatever), in case I have to roll back later on.

In a more technical team, it would be nice to just store the whole thing in git, clone the repository, fire it up locally to create content changes (or add a blog post, etc.), build it locally before committing or triggering a build via something like Netlify, and pulling in and serving the static files on a server via a webhook. That way the team can always keep the source material up to date without requiring too much heavy lifting from a server or exposing the possibility of a security breach of a dynamic interface. 

Not to say that your security is poor, rather that I would just like to limit the attack surface as much as possible. Having flashbacks atm 😅

**Comment by pbuzdin** [comment]
Static builds would be awesome

**Comment by Jack89ita** [comment]
Absolutely a must have feature in my opinion.

**Comment by amiller-gh** [comment]
Any updates on this? Happy to help implement 🙂

**Comment by dwehrmann** [comment]
I agree with @trst and @KeizerDev , this should be mandatory in production. It would be a blast to generate static files and push them to netlify or any other similiar solution. The benefit of giving a user-friendly dashboard/backend to the client along with a opportunity to create a quick static build is the main reason we're looking into "new" cmses right now.

**Comment by heyflo** [comment]
Any updates on this ?
I am also looking forward to this, so I can keep a dashboard locally to edit my client content and push a static site in production.

**Comment by instanceofme** [comment]
This still seems quite popular, maybe don't close it? 😅

**Comment by yougotashovel** [comment]
Any new updates? 
Would love to deploy this to Netlify.

**Comment by amiller-gh** [comment]
👋 Hi all! Just making an announcement that I'll be looping back to this finally in the next few weeks – I have a new project that'll require it! Excited to dive in again 🙂

**Comment by amiller-gh** [comment]
Will continue working in https://github.com/vapid/vapid/pull/99

**Comment by srobbin** [comment]
This just shipped in v0.9.0. To update: `npm install -g @vapid/cli`

A huge thank you to @amiller-gh for adding this feature. It's a great addition to Vapid, and I'm grateful for his hard work.
